18AL: Add support for buying companies between players

perwestling opened the pull request:
This feature makes it possible for a player to buy a private company from another player during the stock round.

This should maybe be a draft instead?

There are two things remaining:
- The layout did not look good if a player buys the first company of two. So need help to get that correct.
- Need to fit this buy together with other so a player cannot buy more than one thing during an SR

perwestling commented:
This is how it looks before and after you click on the top company.
So, no good.

![Screenshot from 2020-08-06 00-53-11](https://user-images.githubusercontent.com/2631151/89471880-84f9c000-d77f-11ea-93f9-179df595ada1.png)
![Screenshot from 2020-08-06 00-53-33](https://user-images.githubusercontent.com/2631151/89471890-888d4700-d77f-11ea-9c59-3848d30876dc.png)

tobymao commented:
Does this even need a separate step? Can't we just modify the original buy companies step to support this?

tobymao commented:
i don't think the button should be here, i think it will be easier for the ui if when you click on a company, it creates selected_company

and then you just put the button below the company to click buy

perwestling commented:
Improved the commit:
- Now the restriction of just one buy per SR does include the companies, so at max a player can buy one share or one company, not both
- The entry field and the button have now been moved to between player cards and the bank saldo / stockmarket

Remaining problem:
- Selection of companies in the player cards does not propagate to the BuyCompanyFromOtherPlayer component. Attempt was made to introduce a :selected_companies, which is updated when a company is selected to show description, but this does not seem to propagate from the Company component

perwestling commented:
Did not get the selection of company to propagate. And it was possibly a bit messy as you could select multiple companies, and also select companies owned by the corporation.

So instead I choose to present a line with one button for each purchasable company. But the line is only shown if the player has not yet bought anything this stock turn.

The button row is below the player cards, so it is similar to when happens when you click on a corporation.

Also, it is a bit down as it is a very rare action, and we much more often want to access the player information.

So I think this can be a Good Enough (TM) solution that has very little impact outside of 18AL. If more 18xx have this functionality, we can consider to make it more general.

![Screenshot from 2020-08-06 14-56-07](https://user-images.githubusercontent.com/2631151/89534589-777f1d00-d7f5-11ea-8ddb-067ec00efabc.png)

perwestling commented:
> Does this even need a separate step? Can't we just modify the original buy companies step to support this?

It is of course possible to modify the original step. The main differences is in process_buy_company. 
But keeping it in a step under g_18_al means it does not complicate the main code with special cases.
When a similar functionality is needed for another 18XX we can think about refactoring it to incorporate this into the common code.

tobymao commented on assets/app/view/game/buy_company_from_other_player.rb:
```diff
@@ -0,0 +1,64 @@
+    class BuyCompanyFromOtherPlayer < Snabberb::Component
+      include Actionable
+
+      needs :game, store: true
```
do not list this as store since you're passing it in

perwestling replied:
Fixed

tobymao commented on assets/app/view/game/buy_company_from_other_player.rb:
```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
```
need to have require actionable to be safe here

perwestling replied:
Fixed

tobymao commented on assets/app/view/game/buy_company_from_other_player.rb:
```diff
@@ -0,0 +1,64 @@
+      include Actionable
+
+      needs :game, store: true
+      needs :selected_companies, default: [], store: true
```
why do you need selected_companies?

perwestling replied:
Removed in latest version

tobymao commented on assets/app/view/game/buy_company_from_other_player.rb:
```diff
@@ -0,0 +1,64 @@
+      def render
+        @step = @game.round.active_step
+        @current_actions = @step.current_actions
+        return h(:div, {}, '') if !@step.can_buy_any?(@game.current_entity) || !@current_actions.include?('buy_company')
```
can just do h(:div) i think? or maybe h(:div, '')

perwestling replied:
Need for this removed in later revision

tobymao commented on assets/app/view/game/buy_company_from_other_player.rb:
```diff
@@ -0,0 +1,64 @@
+        })
+        children = [input]
+        buyable_companies
+          .select { |c| purchasable?(c) }
```
why are you doing select purchasable twice?

perwestling replied:
Ops. Fixed

tobymao commented on assets/app/view/game/buy_company_from_other_player.rb:
```diff
@@ -0,0 +1,64 @@
+          value: 1,
+          type: 'number',
+          min: 1,
+          max: @game.current_entity.cash.to_s.size,
```
maybe store this as a var

perwestling replied:
Fixed

tobymao commented on assets/app/view/game/buy_company_from_other_player.rb:
```diff
@@ -0,0 +1,64 @@
+
+      def purchasable?(company)
+        @game.class::COMPANY_BUY_BETWEEN_PLAYERS &&
+        company&.owner&.player? &&
```
why would company be nil?

perwestling replied:
Hm, fixed

tobymao commented on assets/app/view/game/buy_company_from_other_player.rb:
```diff
@@ -0,0 +1,64 @@
+
+      def render_company_buy(input, company)
+        buy_company = lambda do
+          price = input.JS['elm'].JS['value'].to_i
```
you can do Native(input).elm.value.to_i

perwestling replied:
Fixed

tobymao commented on lib/engine/step/g_18_al/buy_sell_par_shares.rb:
```diff
@@ -6,11 +6,19 @@ module Engine
 
+        def actions(entity)
+          actions = super
+          actions << 'buy_company' if actions.include?('buy_shares')
```
hm, shouldn't you only add buy_company action if there are companies to buy? then you wouldn't need to do all those checks in the view that you do

perwestling replied:
Rewritten this quite a bit, but now the buy_company will only be available if there is a legal buy.

tobymao commented on assets/app/view/game/round/stock.rb:
```diff
@@ -46,6 +46,7 @@ def render
 
           children.concat(render_corporations)
           children << h(Players, game: @game)
+          children << h(BuyCompanyFromOtherPlayer, game: @game) if @game.class::COMPANY_BUY_BETWEEN_PLAYERS
```
you can get rid of all those return h(:div, '') if the actions returns buy company, it will be cleaner

perwestling replied:
Yes, it became much cleaner after applying your suggestion

perwestling commented:
Have now incorporated review comments by @tobymao 

Did a rebase and force push as I rewrote this a bit. Now use the event status feature.
Also have moved the handling of buy_company into the general buy_sell_par_shares class.
Was not able to get adding buy_company to the stock round.

In this case I think buy_company belongs to the buy_sell_par_shares as otherwise we need to introduce a state that would be shared between buy_sell_par_shares and buy_company (@current_actions)

Even if doing that, using the normal buy_company did introduce some special things. Think I needed to check if OR or SR at some places.

Anyway I did not get it to work, but when I tried I think the code was not fully correct yet. So should I give it a new try to make buy_company work for both OR and SR? Introducing a shared round state between buy_company and buy_sell_par_shares? (The shared state is to keep track if anything has been bought during the SR.)

tobymao commented on lib/engine/step/buy_sell_par_shares.rb:
```diff
@@ -188,6 +189,17 @@ def can_ipo_any?(entity)
+        @game.purchasable_companies.reject { |c| c.owner == entity }
+      end
+
+      def can_buy_any_company?(entity)
```
i think this is not necessary you can just put this all in purchasebl_companies

perwestling replied:
Done

tobymao commented on lib/engine/step/buy_sell_par_shares.rb:
```diff
@@ -188,6 +189,17 @@ def can_ipo_any?(entity)
         !bought? && @game.corporations.any? { |c| c.can_par?(entity) && can_buy?(entity, c.shares.first&.to_bundle) }
       end
 
+      def purchasable_companies_from_others(entity)
```
return [] if bought? || !entity.cash.positive?...

@game.purchasable_companies.reject...

this is how you can consolidate can_buy_any_company?

perwestling replied:
Have made an attempt now. Using name purchasable_companies in the step.

tobymao commented:
> Have now incorporated review comments by @tobymao
> 
> Did a rebase and force push as I rewrote this a bit. Now use the event status feature.
> Also have moved the handling of buy_company into the general buy_sell_par_shares class.
> Was not able to get adding buy_company to the stock round.
> 
> In this case I think buy_company belongs to the buy_sell_par_shares as otherwise we need to introduce a state that would be shared between buy_sell_par_shares and buy_company (@current_actions)
> 
> Even if doing that, using the normal buy_company did introduce some special things. Think I needed to check if OR or SR at some places.
> 
> Anyway I did not get it to work, but when I tried I think the code was not fully correct yet. So should I give it a new try to make buy_company work for both OR and SR? Introducing a shared round state between buy_company and buy_sell_par_shares? (The shared state is to keep track if anything has been bought during the SR.)

so is this pr working or missing something? it seems like you're saying it doesn't work in the OR?

perwestling commented:
> so is this pr working or missing something? it seems like you're saying it doesn't work in the OR?

It does work and passes the rake test for 18AL (which has added two purchases in SR1).

What does not (or rather did not) work is (was) to use buy_company as a step in StockRound.

So the question is if I should make another try? It might complicate the buy_company code.

An alternative could be to extract the process_buy_company code to a shared module, which then buy_company and buy_sell_par_shares could share.

perwestling commented:
It should be noted that this PR will break existing alpha 18AL games, as players will be needing to pass during SR. If a player have $1 up to less than cheapest share, in SR1, they before were passed automatically but now have to manually pass, as there are purchasable privates. 

But this is the last needed feature to make 18AL feature complete, and maybe ready for Beta.
